src/shared/bodies/base_body.h

```diff
@@ -35,191 +35,185 @@
-		void defineAdaptation(ConstructorArgs &&...args)
-		{
-			sph_adaptation_ = sph_adaptation_ptr_keeper_
-								  .createPtr<AdaptationType>(*this, std::forward<ConstructorArgs>(args)...);
-		};
-
-		template <typename... ConstructorArgs>
-		LevelSetShape *defineComponentLevelSetShape(const std::string &shape_name, ConstructorArgs &&...args)
-		{
-			ComplexShape *complex_shape = DynamicCast<ComplexShape>(this, body_shape_);
-			return complex_shape->defineLevelSetShape(*this, shape_name, std::forward<ConstructorArgs>(args)...);
-		};
-
-		template <typename... ConstructorArgs>
-		LevelSetShape *defineBodyLevelSetShape(ConstructorArgs &&...args)
-		{
-			LevelSetShape *levelset_shape =
-				shape_ptr_keeper_.resetPtr<LevelSetShape>(*this, *body_shape_, std::forward<ConstructorArgs>(args)...);
-
-			body_shape_ = levelset_shape;
-			return levelset_shape;
-		};
-
-		/** partial construct particles with an already constructed material */
-		template <class ParticleType = BaseParticles, class MaterialType = BaseMaterial>
-		void defineParticlesWithMaterial(MaterialType *material)
-		{
-			base_material_ = material;
-			base_particles_ = base_particles_ptr_keeper_.createPtr<ParticleType>(*this, material);
-		};
-
-		/** partial construct particles with material informaiton. note that particle data not initialized yet */
-		template <class ParticleType = BaseParticles, class MaterialType = BaseMaterial, typename... ConstructorArgs>
-		void defineParticlesAndMaterial(ConstructorArgs &&...args)
-		{
-			MaterialType *material = base_material_ptr_keeper_.createPtr<MaterialType>(std::forward<ConstructorArgs>(args)...);
-			defineParticlesWithMaterial<ParticleType>(material);
-		};
-
-		/** initialize particle data using a particle generator for geometric data.
-		 * the local material parameters are also initialized. */
-		template <class ParticleGeneratorType, typename... ConstructorArgs>
-		void generateParticles(ConstructorArgs &&...args)
-		{
-			sph_adaptation_->registerAdaptationVariables(*base_particles_);
-			ParticleGeneratorType particle_generator(*this, std::forward<ConstructorArgs>(args)...);
-			particle_generator.generateParticlesWithBasicVariables();
-			base_particles_->initializeOtherVariables();
-			base_material_->initializeLocalParameters(base_particles_);
-		};
-
-		template <typename VariableType>
-		void addBodyStateForRecording(const std::string &variable_name)
-		{
-			base_particles_->template addVariableToWrite<VariableType>(variable_name);
-		};
-
-		template <class DerivedVariableMethod,typename... Args>
-		void addDerivedBodyStateForRecording(Args&&... args)
-		{
-			base_particles_->template addDerivedVariableToWrite<DerivedVariableMethod>(std::forward<Args>(args)...);
-		};
-
-        template <class DerivedVariableMethod,typename... Args>
-		void addDerivedBodyState(Args&&... args)
-		{
-			base_particles_->template addDerivedVariable<DerivedVariableMethod>(std::forward<Args>(args)...);
-		};
-
```

Comment:
What is the better way to add a derived variable after this PR, directly adding it to the BaseParticles ?

Reply:
Yes. As data output is done in base particles.